src/session_concurrency/response_json_utils.cc

```diff
@@ -0,0 +1,91 @@
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "response_json_utils.h"
+
+#include <rapidjson/document.h>
+#include <rapidjson/error/error.h>
+
+#include <cstdint>
+#include <stdexcept>
+#include <vector>
+
+namespace triton::perfanalyzer {
+
+const rapidjson::Value&
+ResponseJsonUtils::GetResponseMessage(
+    const std::vector<uint8_t>& response_buffer)
+{
+  const auto response_document{GetResponseDocument(response_buffer)};
+  return GetMessage(response_document);
+}
+
+const rapidjson::Document
+ResponseJsonUtils::GetResponseDocument(
+    const std::vector<uint8_t>& response_buffer)
+{
+  rapidjson::Document response_document{};
+
+  const char* response_buffer_c_str{
+      reinterpret_cast<const char*>(response_buffer.data())};
+
+  response_document.Parse(response_buffer_c_str, response_buffer.size());
+
+  if (response_document.HasParseError()) {
+    throw std::runtime_error(
+        "rapidjson parse error " +
+        std::to_string(response_document.GetParseError()));
+  }
+
+  return response_document;
+}
+
+const rapidjson::Value&
+ResponseJsonUtils::GetMessage(const rapidjson::Document& response_document)
+{
+  if (!response_document.IsObject() ||
+      !response_document.HasMember("choices") ||
+      !response_document["choices"].IsArray() ||
+      response_document["choices"].Empty()) {
+    throw std::runtime_error(
+        "Response body must be an object and have a 'choices' field that is "
+        "an array with at least one element.");
+  }
+
+  const auto& response_first_choice{response_document["choices"][0]};
+
+  if (!response_first_choice.IsObject() ||
```

Comment:
Would it be useful to print the response_first_choice? This message might not give enough information, especially if it's the response that's causing the issue.

Reply:
Done 👍